Automattic/woocommerce-payments
# Add platform checkout signature endpoint

#### Changes proposed in this Pull Request


Add an endpoint to get platform checkout signature


After #4785 added a signature to the user exists endpoint, a follow up issue was created to create an endpoint to get this signature so that it could be retrieved when needed as the signature is time sensitive. The issue is that the signature can become stale if the user is on the checkout page for too long, so this solution is to add a way to retrieve the signature when it is needed so that there isn't any risk to it being stale.

There were two options to approach this, one was to retrieve the signature on an interval and the other was to retrieve it once it was needed. I opted to retrieve it once it was needed as a way to eliminate (possibly many) unnecessary requests to the merchant store as well as avoid some complexity with retrieving it at an interval. The downside to this approach is a slightly longer time to check the `users/exists` endpoint on the platform checkout. This seemed like a reasonable trade off to me.

The changes are straightforward and use the WC Ajax so that a nonce can be verified in the right context.





#### Testing instructions



- Checkout this branch locally with a working WooPay setup.
- Set up cart with an item and go to checkout.
- Fill out the email address field on the shortcode checkout.
- Confirm that there was a request to the merchant store at: `?wc-ajax=wcpay_get_platform_checkout_signature`.
- Confirm the request included an `_ajax_nonce`.
- Confirm the response was successful and contained a signature.
- Confirm that signature was then passed as a query parameter to the WooPay `user/exists` endpoint.
- Confirm that the user exists look up responded successfully (whether true or false).

**Test some failing requests**

- Modify line https://github.com/Automattic/woocommerce-payments/compare/add/1313-platform-checkout-signature-endpoint?expand=1#diff-d65a9401ef14963cb2c9f8d930e98c69f81e5832f50fe794934471c4fd1a8c4fR340 to have an incorrect AJAX URL (for example, I changed it to: `get_platform_checkout_signatureAAA`) and refresh the checkout page.
- Confirm that the request does not return successfully.
- Confirm that the `WooPay is unavailable...` error message is displayed below the email field.
- Confirm the spinner was removed.
- Confirm that there was NO request to the WooPay `user/exists` endpoint.
- Fix the previous typo in the AJAX action.
- Append some characters to the `_ajax_nonce` to simulate it failing and refresh the checkout page.
- Confirm that the request returns a 403.
- Confirm that the `WooPay is unavailable...` error message is displayed below the email field.
- Confirm the spinner was removed.
- Confirm that there was NO request to the WooPay `user/exists` endpoint.



-------------------


**Post merge**
